shell/platform/android/io/flutter/embedding/android/FlutterActivity.java

```diff
@@ -389,6 +391,20 @@ public FlutterActivity() {
     lifecycle = new LifecycleRegistry(this);
   }
 
+  /**
+   * This method exists so that JVM tests can ensure that a delegate exists without
+   * putting this Activity through any lifecycle events, because JVM tests cannot handle
+   * executing any lifecycle methods, at the time of writing this.
+   * <p>
+   * The testing infrastructure should be upgraded to make FlutterActivity tests easy to
+   * write while exercising real lifecycle methods. At such a time, this method should be
+   * removed.
+   */
+  @VisibleForTesting
+  /* package */ void setDelegate(@NonNull FlutterActivityAndFragmentDelegate delegate) {
```

Comment:
Being able to put in a delegate factory was the reason why I made the FlutterInjector :D

Reply:
Yeah, I think the desired outcome here is to better support common testing mechanisms. Then we would probably have alternative approaches that don't require this or an injector.

Reply:
+1. We don't necessarily need to bring in a full DI framework (and probably shouldn't) but it would help testing a ton in general if the classes used constructor injection as a pattern instead of defaulting to instantiating everything as silent side effects of being constructed, or having certain callbacks triggered.

Reply:
Yeah, the combination of problems here is that:

- We can't constructor inject framework classes
- The very 1st lifecycle method in FlutterActivity and FlutterFragment attempt to connect to JNI

We should eventually probably introduce a static dependency resolver for `FlutterActivity` and `FlutterFragment`. @xster had something like that in his original PR. The reason I removed it was because I think it was also being passed to plain Java classes, which I think should always utilize constructor injection unless there is an exceptional situation. But I probably should not have removed it entirely - I should have left it in specifically for `FlutterActivity` and `FlutterFragment`

At this point my recommendation is that we do what is necessary to button up the embedding for public consumption, and then it would be nice to schedule a dedicated project to resolve outstanding issues related to testing including: testing dep resolution, testing dep versioning, and embedding tools for testing.

Reply:
ya, this is now tangential to this PR to some extent. But ya, that's why DI is so common on Android, because you can't construct and inject stuff generally.

Reply:
Let's clean these up at some point in your TODO issue.